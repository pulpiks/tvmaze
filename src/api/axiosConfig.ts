import axios from 'axios'
import { BASE_URL_API, TIMEOUT } from '../constants/config'
import extendAxiosError from '../utils/errorHandler'

const axiosInstance = axios.create({
  baseURL: BASE_URL_API,
  timeout: TIMEOUT
})

axiosInstance.interceptors.request.use(
  (config: any) => config,
  (error: any) => Promise.reject(error)
)

axiosInstance.interceptors.response.use(
  (response: any) => response,
  (error: any) => Promise.reject(extendAxiosError(error))
)

export default axiosInstance
