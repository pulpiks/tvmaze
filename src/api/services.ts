import { SHOWS_ALL_API, SEARCH_SHOWS_BY_NAME } from '@/constants/endpoints'
import axiosInstance from './axiosConfig'

export const servicesAPI = {
  getShows: () => axiosInstance.get(SHOWS_ALL_API),
  searchByText: (q: string) => axiosInstance.get(`${SEARCH_SHOWS_BY_NAME}?q=${q}`),
  getShowById: (id: string) => axiosInstance.get(`${SHOWS_ALL_API}/${id}`)
}
