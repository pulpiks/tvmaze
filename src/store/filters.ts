export enum TEnumSortRating {
  ASC = 'ASC',
  DESC = 'DESC',
}

export enum TEnumFilters {
  genres = 'genres',
  sortRating = 'sortRating',
  searchText = 'searchText'
}

export type TFilterState = {
  [TEnumFilters.genres]: string[],
  [TEnumFilters.sortRating]: TEnumSortRating,
  [TEnumFilters.searchText]: string
}

export const defaultFiltersState: TFilterState = {
  [TEnumFilters.genres]: [],
  [TEnumFilters.sortRating]: TEnumSortRating.DESC,
  [TEnumFilters.searchText]: ''
}

export const SORTING = {
  [TEnumSortRating.ASC]: 'lowest rated',
  [TEnumSortRating.DESC]: 'top rated'
}
