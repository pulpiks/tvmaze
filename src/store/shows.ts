export type TGenre = string;

export enum TStatusEnum {
  Ended = 'Ended',
  Running = 'Running',
  ToBeDetermined = 'To Be Deterrmined'
}

type TDay = 'Monday' | 'Tuesday' | 'Wednesday' | 'Thursday' | 'Friday' | 'Saturday' | 'Sunday';

type TSchedule = {
  time: string;
  days: TDay[];
};

type TRating = {
  average?: number;
};

type TNetwork = {
  id: number;
  name: string;
  country: {
    name: string;
    code: string;
    timezone: string;
  };
};

type TWebChannel = {
  id: number;
  name: string;
  country: string | null;
} | null;

type TImage = {
  medium: string;
  original: string;
};

export type TShow = {
  id: number;
  url: string;
  name: string;
  type: string;
  language: string;
  genres: TGenre[];
  status: TStatusEnum;
  runtime: number;
  premiered: string;
  officialSite: string;
  schedule: TSchedule;
  rating: TRating;
  weight: number;
  network: TNetwork;
  webChannel: TWebChannel;
  externals: {
    tvrage: number;
    thetvdb: number;
    imdb: string;
  };
  image: TImage;
  summary: string;
  updated: number;
  _links: {
    self: {
      href: string;
    };
    previousepisode?: {
      href: string;
    };
  };
};

export type TShowState = TShow[]

export const defaultShowsState: TShowState = []
export const FETCH_SHOWS = 'FETCH_SHOWS'
export const FETCH_SHOWS_BY_SEARCH = 'FETCH_SHOWS_BY_SEARCH'
