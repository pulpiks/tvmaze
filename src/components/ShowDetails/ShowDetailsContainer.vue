<template>
  <div class="show-page" data-test="show-page">
    <b-container v-if="show" fluid data-test="show">
      <div v-if="loading" class="d-flex justify-content-center my-2">
        <b-spinner label="Loading..." variant="secondary"></b-spinner>
      </div>
      <div v-else>
        <show-details :show="show"></show-details>
      </div>
    </b-container>
  </div>
</template>

<script lang="ts">
import { servicesAPI } from '@/api/services'
import { generalErrorHandler } from '@/utils/errorHandler'
import { Component, Vue, Prop } from 'vue-property-decorator'
import ShowDetails from './ShowDetails.vue'

@Component({
  name: 'ShowDetailsContainer',
  components: {
    ShowDetails
  }
})

export default class extends Vue {
  private loading = true

  private show = null

  get showId() {
    return this.$route.params.id
  }

  created() {
    this.fetchShow()
  }

  async fetchShow() {
    try {
      this.loading = true
      const { data: show } = await servicesAPI.getShowById(this.showId)
      this.show = show
    } catch (err) {
      generalErrorHandler(this)(err)
    } finally {
      this.loading = false
    }
  }
}
</script>
