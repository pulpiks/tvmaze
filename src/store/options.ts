import { StoreOptions } from 'vuex'
import VuexPersistence from 'vuex-persist'

import {
  CHANGE_FILTERS, FILTER_GENRE, FILTER_RATING, RESET_SEARCH, FILTER_SEARCH
} from '@/components/ShowsList/ShowsListTypes'
import {
  TShowState, defaultShowsState, FETCH_SHOWS, TShow, FETCH_SHOWS_BY_SEARCH
} from './shows'
import { TFilterState, defaultFiltersState } from './filters'

export type S = {
  shows: TShowState,
  filters: TFilterState,
}

const initialState = {
  shows: defaultShowsState,
  filters: defaultFiltersState
}

const vuexLocal = new VuexPersistence({
  key: 'tvmaze',
  storage: window.localStorage,
  reducer: (state: S) => ({
    filters: {
      ...state.filters
    }
  })
})


export const options: StoreOptions<any> = {
  plugins: [vuexLocal.plugin],
  state: { ...initialState },
  mutations: {
    [FETCH_SHOWS](state, action) {
      state.shows = action.payload
    },

    [RESET_SEARCH](state) {
      state.filters[FILTER_SEARCH] = ''
    },
    // fetchServiceDetails(state, action) {
    //   state.serviceDetails = action.payload
    //   state.currentServiceVersion = null
    // },
    [CHANGE_FILTERS](state, action) {
      state.filters[action.filter] = action.values
    }
    // [FETCH_SHOWS_BY_SEARCH](state, action) {
    //   state.shows = action.payload
    // }
  },
  actions: {
    [FETCH_SHOWS]({ commit }, action) {
      commit(action)
    },
    [CHANGE_FILTERS]({ commit }, action) {
      const { payload } = action
      switch (payload.filter) {
        case FILTER_GENRE:
          commit(CHANGE_FILTERS, payload)
          break
        case FILTER_RATING:
          commit(CHANGE_FILTERS, payload)
          break
        case FILTER_SEARCH:
          commit(CHANGE_FILTERS, payload)
          break
        default:
          break
      }
    },
    // [FETCH_SHOWS_BY_SEARCH]({ commit }, action) {
    //   commit(FETCH_SHOWS_BY_SEARCH, action)
    // },
    [RESET_SEARCH]({ commit }) {
      commit(RESET_SEARCH)
    }
    // fetchServiceDetails({ commit }, payload) {
    //   commit(payload)
    // },
    // changeFilter({ commit }, action) {
    //   const { payload } = action
    //   switch (payload.filter) {
    //     case 'types':
    //       commit('changeFilter', payload)
    //       break
    //     case 'technicalTypes':
    //       commit('changeFilter', payload)
    //       break
    //     case 'statuses':
    //       commit('changeFilter', payload)
    //       break
    //     default:
    //       break
    //   }
    // },
    // changeVersion({ commit }, payload) {
    //   commit(payload)
    // },
    // resetAllFilters({ commit }) {
    //   commit('resetAllFilters')
    // },
    // clearServiceDetails({ commit }) {
    //   commit('clearServiceDetails')
    // }
  },
  getters: {
    allGenres: state => state.shows.reduce((genres: string[], curShow: TShow) => {
      curShow.genres.forEach((g) => {
        if (!genres.includes(g)) {
          genres.push(g)
        }
      })
      return genres
    }, [])

    // filteredShows: state => state.shows
    //   .filter(show => show.genres.find(g => state.getters.genres.has(g)))


    // searchText: state => state.servicesList.searchText,
    // servicesFilters: state => state.servicesList,
    // serviceDetails: state => state.serviceDetails,
    // areAnyFiltersActive: state => {
    //   const {
    //     searchText,
    //     filters: { types, technicalTypes, statuses }
    //   } = state.servicesList
    //   return searchText || types.length || technicalTypes.length || statuses
    // }
  }
}
