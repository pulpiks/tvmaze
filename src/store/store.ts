import Vue from 'vue'
import Vuex from 'vuex'
import { options, S } from './options'


Vue.use(Vuex)

export default new Vuex.Store<S>(options)
