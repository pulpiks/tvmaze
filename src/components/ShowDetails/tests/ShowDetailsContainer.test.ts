import { createLocalVue, shallowMount } from '@vue/test-utils'
import { BootstrapVue, BootstrapVueIcons } from 'bootstrap-vue'
import ShowDetailsContainer from '../ShowDetailsContainer.vue'
import store from '../../../store/store'
import testData from './testData'

const mockSuccessResponse = testData

jest.mock('../../../api/services', () => {
  const services = jest.requireActual('../../../api/services')
  return {
    ...services,
    servicesAPI: {
      ...services.servicesAPI,
      getShowById: jest.fn()
    }
  }
})

// eslint-disable-next-line import/first
import { servicesAPI } from '../../../api/services'

describe('ShowDetailsContainer', () => {
  let localVue: any
  beforeEach(() => {
    localVue = createLocalVue()
    localVue.use(BootstrapVue)
    localVue.use(BootstrapVueIcons)
  })

  afterEach(() => {
    jest.resetAllMocks()
  })

  it('renders correctly', async (done) => {
    (servicesAPI.getShowById as unknown as jest.Mock).mockImplementation(() => Promise.resolve({
      data: mockSuccessResponse
    }))

    const $route = {
      params: {
        id: '1'
      }
    }

    const wrapper = await shallowMount(ShowDetailsContainer, {
      localVue,
      store,
      mocks: {
        $route
      }
    })

    wrapper.vm.$nextTick(() => {
      expect(wrapper.element).toMatchSnapshot()
      expect(wrapper.find('[data-test=\'show-page-details\']').exists()).toBe(true)
      done()
    })
  })
})
