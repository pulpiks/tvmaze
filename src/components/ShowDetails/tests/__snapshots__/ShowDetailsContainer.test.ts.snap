// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShowDetailsContainer renders correctly 1`] = `
<div
  class="show-page"
  data-test="show-page"
>
  <b-container-stub
    data-test="show-page-container"
    fluid="true"
    tag="div"
  >
    <div>
      <show-details-stub
        data-test="show-page-details"
        show="[object Object]"
      />
    </div>
  </b-container-stub>
</div>
`;
